core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java

```diff
@@ -63,22 +63,29 @@ public Result compute(
     TokenFactory tokenFactory = null;
 
     Map<UUID, DefaultNode> newNodes = new HashMap<>();
+    // Contact point nodes don't have host ID as well as other info yet, so we fill them with node
+    // info found on first match by endpoint
+    Set<DefaultNode> matchedContactPoints = new HashSet<>();
```

Comment:
I'm a bit surprised: `DefaultNode` doesn't implement `hashCode()` and `equals()`. I wonder if methods like `contains()` and `add()` (used below) won't misbehave? Wouldn't be better to use `Set<EndPoint>` here? It seems doable.

Reply:
ah right interesting, didn't occur to me at the time, but made a similar observation on the (initial commit)[https://github.com/apache/cassandra-java-driver/pull/2007#discussion_r1934067152].  I suspect that for some reason this still works as I still see `Adding new node {}` in my test, double checking this though.

Reply:
Oh right, because `findContactPointNode` grabs the `Node` from `contactPoints` and adds it to `matchedContactPoints` .  On each iteration `findContactPointNode` grabs the exact object, so `matchedContactPoints.contains(contactPointNode)` evaluates to true after the first encountered one.

Reply:
Yeah that's what I suspected. It only works because the objects are the same. But still, that's a bit fragile. I think using `Set<EndPoint>` would also work, and would be more future-proof.

Reply:
agreed :+1:

Reply:
Hi @adutra 👋 .
Thanks for the review.
I actually relied on equality of references here, so that was intentional. I can look into refactoring to use EndPoint, from the first PoV it should work same. If we assume that contact points are always linked to the different EndPoints all will be smooth.

Reply:
I see we store contact point EndPoints in Set when creating Nodes for them in `com.datastax.oss.driver.internal.core.metadata.MetadataManager#addContactPoints`, so we should be good to assume their uniqueness.

Reply:
Pushed a commit to address the proposal, @adutra please take a look 🙏 .
If you approve, I will squash the commits and add the required commit description about the author and reviewers (linked it in the footer of the PR description).